File: 54-spiral-matrix/spiral-matrix.cs
public class Solution {
    public IList<int> SpiralOrder(int[][] matrix) {
        var result = new List<int>();
        if (matrix == null || matrix.Length == 0 || matrix[0].Length == 0)
            return result;

        int top = 0, left = 0;
        int bottom = matrix.Length - 1, right = matrix[0].Length - 1;

        while (top <= bottom && left <= right) {
            for (int i = left; i <= right; i++) 
                result.Add(matrix[top][i]);
            top++;

            for (int i = top; i <= bottom; i++) 
                result.Add(matrix[i][right]);
            right--;

            if (top <= bottom) {
                for (int i = right; i >= left; i--) 
                    result.Add(matrix[bottom][i]);
                bottom--;
            }

            if (left <= right) {
                for (int i = bottom; i >= top; i--) 
                    result.Add(matrix[i][left]);
                left++;
            }
        }

        return result;
    }
}

File: 54-spiral-matrix/spiral-matrix.ts
function spiralOrder(matrix: number[][]): number[] {
    const result: number[] = [];
    if (!matrix || matrix.length === 0 || matrix[0].length === 0)
        return result;

    let top = 0, left = 0;
    let bottom = matrix.length - 1, right = matrix[0].length - 1;

    while (top <= bottom && left <= right) {
        for (let i = left; i <= right; i++) 
            result.push(matrix[top][i]);
        top++;

        for (let i = top; i <= bottom; i++) 
            result.push(matrix[i][right]);
        right--;

        if (top <= bottom) {
            for (let i = right; i >= left; i--) 
                result.push(matrix[bottom][i]);
            bottom--;
        }

        if (left <= right) {
            for (let i = bottom; i >= top; i--) 
                result.push(matrix[i][left]);
            left++;
        }
    }

    return result;
}
